test/e2e/e2e_leaderelection_test.go

```diff
@@ -0,0 +1,195 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"reflect"
+	"strings"
+	"testing"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	clientset "k8s.io/client-go/kubernetes"
+
+	"sigs.k8s.io/descheduler/cmd/descheduler/app/options"
+	"sigs.k8s.io/descheduler/pkg/descheduler"
+)
+
+func TestLeaderElection(t *testing.T) {
+	ctx := context.Background()
+
+	clientSet, _, _, stopCh := initializeClient(t)
+	defer close(stopCh)
+
+	ns1 := "e2e-" + strings.ToLower(t.Name()+"-a")
+	ns2 := "e2e-" + strings.ToLower(t.Name()+"-b")
+
+	t.Logf("Creating testing namespace %v", ns1)
+	testNamespace1 := &v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: ns1}}
+	if _, err := clientSet.CoreV1().Namespaces().Create(ctx, testNamespace1, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Unable to create ns %v", testNamespace1.Name)
+	}
+	defer clientSet.CoreV1().Namespaces().Delete(ctx, testNamespace1.Name, metav1.DeleteOptions{})
+
+	t.Logf("Creating testing namespace %v", ns2)
+	testNamespace2 := &v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: ns2}}
+	if _, err := clientSet.CoreV1().Namespaces().Create(ctx, testNamespace2, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Unable to create ns %v", testNamespace2.Name)
+	}
+	defer clientSet.CoreV1().Namespaces().Delete(ctx, testNamespace2.Name, metav1.DeleteOptions{})
+
+	deployment1, err := createDeployment(ctx, clientSet, ns1, 5, t)
+	if err != nil {
+		t.Fatalf("create deployment 1: %v", err)
+	}
+	defer clientSet.AppsV1().Deployments(deployment1.Namespace).Delete(ctx, deployment1.Name, metav1.DeleteOptions{})
+
+	deployment2, err := createDeployment(ctx, clientSet, ns2, 5, t)
+	if err != nil {
+		t.Fatalf("create deployment 2: %v", err)
+	}
+	defer clientSet.AppsV1().Deployments(deployment2.Namespace).Delete(ctx, deployment2.Name, metav1.DeleteOptions{})
+
+	waitForPodsRunning(ctx, t, clientSet, map[string]string{"test": "leaderelection", "name": "test-leaderelection"}, 5, ns1)
+
+	podListAOrg := getPodNameList(ctx, clientSet, ns1, t)
+
+	waitForPodsRunning(ctx, t, clientSet, map[string]string{"test": "leaderelection", "name": "test-leaderelection"}, 5, ns2)
+
+	podListBOrg := getPodNameList(ctx, clientSet, ns2, t)
+
+	s, err := options.NewDeschedulerServer()
+	if err != nil {
+		t.Fatalf("unable to initialize server: %v", err)
+	}
+	s.Client = clientSet
+	s.DeschedulingInterval = 5 * time.Second
+	s.LeaderElection.LeaderElect = true
+	s.KubeconfigFile = os.Getenv("KUBECONFIG")
+
+	s1 := s
```

Comment:
```go
s1, err := options.NewDeschedulerServer()
s2, err := options.NewDeschedulerServer()
```
rather than copying `s` to avoid sharing data.

Reply:
Actually, I did it intentionally to avoid duplication for the following lines:

```go
if err != nil {
		t.Fatalf("unable to initialize server: %v", err)
	}
	s.Client = clientSet
	s.DeschedulingInterval = 5 * time.Second
	s.LeaderElection.LeaderElect = true
	s.KubeconfigFile = os.Getenv("KUBECONFIG")
```

Reply:
It is ok to duplicate those when it comes to the initialization. Other option is to create a function which will produce the same. Invoke twice then.

Reply:
Hi @ingvagabund, I updated the changes as you mentioned.